### PR TITLE
feature(hx): add domainnet dataset

### PR DESCRIPTION
**Pull request by XinHao-96**
昊霖师兄，我刚又仔细对了一下，感觉可能会有一个的问题。
国港师兄的数据集是cross-domain的，所以他之前的加载逻辑是我在main中写的那种调用方式，我看Fling中load-dataset不涉及domain，所以可能要add_argument（而且是不是sample-iid也需要改？）

还请师兄帮我康康，我这边基本把这个数据集的加载逻辑搞明白了，我需要怎么改师兄跟我说下就行！
而且七个数据集的逻辑都差不多，改好一个我就大概i知道别的怎么做了，我需要怎么改师兄跟我说下就行嘿嘿。
辛苦师兄！                               -郝鑫

### CODE REVIEW

**Comment by XinHao-96** [comment]
昊霖师兄！这是我新改的一版，主要的修改我都放在modify_Hao.txt这个文件中做了解释，还请师兄帮我看下！

File: fling/dataset/domainnet.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+import os
+import numpy as np
+from PIL import Image
+from fling.utils.registry_utils import DATASET_REGISTRY
+
+
+@DATASET_REGISTRY.register('domainnet')
+class DomainNetDataset(Dataset):
+    def __init__(self, cfg: dict, domain, train: bool, train_num=105, test_num=-1):
+        super(DomainNetDataset, self).__init__()
```
Comment (kxzxvbk) [inline fling/dataset/domainnet.py:16]: Each function should contain typing lint, for example: "def __init__(self, cfg: dict, domain: str, train: bool, train_num: int=105....)"

File: fling/dataset/domainnet.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+import torchvision.transforms as transforms
+import os
+import numpy as np
+from PIL import Image
+from fling.utils.registry_utils import DATASET_REGISTRY
+
+
+@DATASET_REGISTRY.register('domainnet')
+class DomainNetDataset(Dataset):
+    def __init__(self, cfg: dict, domain, train: bool, train_num=105, test_num=-1):
```
Comment (kxzxvbk) [inline fling/dataset/domainnet.py]: Typically in our standard, train_num and test_num is used when data partition. Is it necessary to use it when data loading?

File: fling/dataset/domainnet.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        if self.train:
+            img, label = self.train_data[index]
+            img = self.transform_train(img)
+        else:
+            img, label = self.test_data[index]
+            img = self.transform_test(img)
+
+        return {'input': img, 'class_id': label}
+
+def load_domainnet(base_dir, domain, train_num=105, test_num=-1):
```
Comment (kxzxvbk) [inline fling/dataset/domainnet.py:54]: Typing lint and comments for this function.

File: fling/pipeline/cross_domain_model_pipeline.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+from fling.component.client import get_client
+from fling.component.server import get_server
+from fling.component.group import get_group
+from fling.dataset import get_cross_domain_dataset
+from fling.utils.data_utils import data_sampling
+from fling.utils import Logger, compile_config, client_sampling, VariableMonitor, LRScheduler, get_launcher
+
+
+def cross_domain_model_pipeline(args: dict, seed: int = 0) -> None:
+    r"""
```
Comment (kxzxvbk) [inline fling/pipeline/cross_domain_model_pipeline.py:13]: Update this comment

**Comment by kxzxvbk** [comment]
Reformat the code style by using ``source format.sh ./``

**Comment by kxzxvbk** [comment]
Add example configs like other datasets (e.g. CIFAR100)

**Comment by XinHao-96** [comment]
师兄我reformat了一下我修改的部分，然后在flzoo中新建了domainnet的一个config，是仿照cifar100做的，麻烦师兄看下：）

**Comment by XinHao-96** [comment]
昊霖师兄，我把cross-domain场景下的一个数据集调通了。现在cross-domain的场景是用了新的pipeline和base-group, client, server, launcher以及Avg的方法，应该是在flzoo里面的domainnet的config是可以直接运行的，结果也和国港师兄对了一下，应该没什么问题，我后续再把其他数据集弄一下。-郝鑫

**Comment by kxzxvbk** [comment]
1) Resolve the conflicts 
2) Merge the main branch
3) Reformat the codes

File: fling/component/client/client_template.py [hunk 5]
```diff
@@ -148,3 +148,147 @@ def test(self) -> dict:
             - A diction containing testing results.
         """
         raise NotImplementedError
+
+class CrossDomainClientTemplate:
+    r"""
+    Overview:
+        Template of client in Federated Learning.
```
Comment (kxzxvbk) [inline fling/component/client/client_template.py:155]: Refine this comment.

File: fling/component/client/client_template.py [hunk 6]
```diff
@@ -148,3 +148,147 @@ def test(self) -> dict:
             - A diction containing testing results.
         """
         raise NotImplementedError
+
+class CrossDomainClientTemplate:
```
Comment (kxzxvbk) [inline fling/component/client/client_template.py:152]: I suppose that the difference between this class and ``ClientTemplate`` only lies in the construction function. So, you can simply inherit ``ClientTemplate`` with minor modifications.

File: fling/component/client/cross_domain_base_client.py [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+from torch.utils.data import DataLoader, Dataset
+import torch.nn as nn
+
+from fling.utils import get_optimizer, VariableMonitor, get_weights
+from fling.utils.registry_utils import CLIENT_REGISTRY
+from .client_template import CrossDomainClientTemplate
+
+
+@CLIENT_REGISTRY.register('cross_domain_base_client')
+class CrossDomainBaseClient(CrossDomainClientTemplate):
```
Comment (kxzxvbk) [inline fling/component/client/cross_domain_base_client.py:14]: Almost the same as ``BaseClient`` ? Inherit the ``BaseClient`` and only do some necessary modifications.

File: fling/component/group/build_group.py [hunk 8]
```diff
@@ -1,7 +1,11 @@
 from fling.utils.registry_utils import GROUP_REGISTRY
 from fling.utils import Logger
 from fling.component.group import ParameterServerGroup
+# from fling.component.group import CrossDomainParameterServerGroup
 
 
 def get_group(args: dict, logger: Logger) -> ParameterServerGroup:
     return GROUP_REGISTRY.build(args.group.name, args, logger)
+
+def get_cross_domain_group(args: dict, logger: Logger):
```
Comment (kxzxvbk) [inline fling/component/group/build_group.py:10]: Why not reuse the ``get_group`` function?

File: fling/component/group/cross_domain_base_group.py [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+
+from fling.utils import get_params_number
+from fling.utils.compress_utils import cross_domain_fed_avg
+from fling.utils.registry_utils import GROUP_REGISTRY
+from fling.utils import Logger, get_weights
+from fling.component.client import CrossDomainClientTemplate
+
+
+@GROUP_REGISTRY.register('cross_domain_base_group')
+class CrossDomainParameterServerGroup:
```
Comment (kxzxvbk) [inline fling/component/group/cross_domain_base_group.py:12]: Inherite the ``ParameterServerGoup``.

File: fling/component/group/cross_domain_base_group.py [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+    r"""
+    Overview:
+        Base implementation of the group in federated learning.
```
Comment (kxzxvbk) [inline fling/component/group/cross_domain_base_group.py:15]: Polish this comment.

File: fling/component/server/cross_domain_base_server.py [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+import torch.nn as nn
+from torch.utils.data import DataLoader, Dataset
+
+from fling.utils import VariableMonitor
+from fling.utils.registry_utils import SERVER_REGISTRY
+from .server_template import ServerTemplate
+
+
+@SERVER_REGISTRY.register('cross_domain_base_server')
+class CrossDomainBaseServer(ServerTemplate):
```
Comment (kxzxvbk) [inline fling/component/server/cross_domain_base_server.py:12]: Inherite the ``BaseServer``

File: fling/dataset/domainnet.py [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+
+
+@DATASET_REGISTRY.register('domainnet')
+class DomainNetDataset(data.Dataset):
+    r"""
+    Implementation for DomainNet dataset. 
+    """
+    def __init__(self, cfg: dict, domain: str,  train: bool): # cfg can be loaded
+        super(DomainNetDataset, self).__init__()
+        self.train_transform = transforms.Compose([
```
Comment (kxzxvbk) [inline fling/dataset/domainnet.py:17]: Do not add data transform like this. Use the default config form. Refer to other dataset files.

File: fling/dataset/domainnet.py [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+        transforms.RandomRotation((-30, 30)),
+        transforms.ToTensor()
+    ])
+        self.test_transform = transforms.Compose([
+        transforms.Resize([256, 256]),
+        transforms.ToTensor()
+    ])
+        self.train = train
+        self.cfg = cfg
+        self.train_imgs, self.train_labels, self.test_imgs, self.test_labels = \
```
Comment (kxzxvbk) [inline fling/dataset/domainnet.py:29]: Load train_imgs and labels when ``train==True`` and otherwise load test_imgs and labels. Do not load both of them simultaneously

File: fling/dataset/domainnet.py [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+        if self.train:
+            return len(self.train_labels)
+        else: 
+            return len(self.test_labels)
+        
+    def __getitem__(self, index: int) -> dict:
+        if self.train:
+            img, label = self.train_imgs[index], self.train_labels[index]
+            if len(img.split()) != 3:
+                img = transforms.Grayscale(num_output_channels=3)(img)
```
Comment (kxzxvbk) [inline fling/dataset/domainnet.py:42]: Add this data transform to ``self.transform``, depending on the name of this domain.

File: fling/dataset/domainnet.py [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+            img = self.train_transform(img)
+        else: 
+            img, label = self.test_imgs[index], self.test_labels[index]
+            if len(img.split()) != 3:
+                img = transforms.Grayscale(num_output_channels=3)(img)
+            img = self.test_transform(img)
+
+
+        # return {'input': img, 'class_id': label}
```
Comment (kxzxvbk) [inline fling/dataset/domainnet.py:51]: Remove comments like this.

File: fling/dataset/domainnet.py [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+
+    if train_num <= len(train_imgs):
+        train_imgs = train_imgs[:train_num]
+        train_labels = train_labels[:train_num]
+
+    if test_num <= len(test_imgs):
+        test_imgs = test_imgs[:test_num]
+        test_labels = test_labels[:test_num]
+
+    print('Load {} Dataset...'.format(domain))
```
Comment (kxzxvbk) [inline fling/dataset/domainnet.py:87]: Don't print this information.

File: fling/pipeline/cross_domain_model_pipeline.py [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+    """
+    # Compile the input arguments first.
+    args = compile_config(args, seed=seed)
+
+    # Construct logger.
+    logger = Logger(args.other.logging_path)
+
+    
+    # Load dataset with domains.
+    # edition: train_set[domain]
```
Comment (kxzxvbk) [inline fling/pipeline/cross_domain_model_pipeline.py:30]: Remove comments like this

File: fling/pipeline/cross_domain_model_pipeline.py [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+            # Get mean results across each client.
+            mean_test_variables = test_monitor.variable_mean()
+
+            # Logging test variables.
+            logger.add_scalars_dict(prefix='after_aggregation_test', dic=mean_test_variables, rnd=i)
+
+            # Saving model checkpoints.
+            torch.save(group.server.glob_dict, os.path.join(args.other.logging_path, 'model.ckpt'))
+
+    # # Fine-tuning
```
Comment (kxzxvbk) [inline fling/pipeline/cross_domain_model_pipeline.py:137]: Polish it

File: fling/utils/compress_utils/fed_avg.py [hunk 19]
```diff
@@ -34,3 +34,41 @@ def fed_avg(clients: list, server: ServerTemplate) -> int:
         trans_cost += len(clients) * state_dict[k].numel()
     # 1B = 32bit
     return 4 * trans_cost
+
+
+def cross_domain_fed_avg(clients: dict, args: dict,  server: ServerTemplate) -> int:
+    r"""
+    Overview:
+        Use the average method to aggregate parameters in different client models.
```
Comment (kxzxvbk) [inline fling/utils/compress_utils/fed_avg.py:42]: Give some more detailed comments on what this function does

File: fling/utils/cross_domain_launcher_utils.py [hunk 20]
```diff
@@ -0,0 +1,153 @@
+from copy import deepcopy
```
Comment (kxzxvbk) [inline fling/utils/cross_domain_launcher_utils.py:1]: Why this file?

File: fling/utils/data_utils/sampling.py [hunk 21]
```diff
@@ -19,6 +20,62 @@ def __len__(self) -> int:
         return len(self.indexes)
 
 
+class DatasetSplit(data.Dataset):
```
Comment (kxzxvbk) [inline fling/utils/data_utils/sampling.py:23]: Remove this class and use NaiveDataset

File: fling/utils/data_utils/sampling.py [hunk 22]
```diff
@@ -19,6 +20,62 @@ def __len__(self) -> int:
...
+
+
+def cross_domain_iid_sampling(datasets: Dataset, num_users: int, train: bool, seed: int):
+    r"""
+    Overview:
+        Independent and identical (i.i.d) sampling method for cross-domain datasets.
+    Arguments:
+        dataset: the total dataset to be sampled (with all domains included)
+        client_number: the number of clients per domain
+        seed: dynamic seed(no need).
```
Comment (kxzxvbk) [inline fling/utils/data_utils/sampling.py:43]: What is the meaning of "no seed"?

File: fling/utils/data_utils/sampling.py [hunk 23]
```diff
@@ -19,6 +20,62 @@ def __len__(self) -> int:
...
+    Returns:
+        A dict of datasets for each client per domain.
+    """
+    def _sampling_iid(dataset: Dataset, num_users: int, seed: int):
+        """
+        Sample I.I.D. client data from dataset
```
Comment (kxzxvbk) [inline fling/utils/data_utils/sampling.py:49]: Polish the format of this comment.

File: flzoo/default_config.py [hunk 24]
```diff
@@ -7,6 +7,12 @@
         dataset='cifar10',
         # Root path for dataset.
         data_path='./data',
+
```
Comment (kxzxvbk) [inline flzoo/default_config.py:10]: 1) Do not modify the default config
2) Do not include the absolute path of your computer. For example, you can put the data to `./data/domain_net`

File: flzoo/domainnet/domainnet_fedavg_resnet_config.py [hunk 25]
```diff
@@ -0,0 +1,35 @@
+from easydict import EasyDict
+
+exp_args = dict(
+    data=dict(
+        dataset='domainnet',
+        data_path='./data/domainnet',
+        sample_method=dict(name='cross_domain_iid', train_num=500, test_num=30000),
+        domains='clipart,infograph,painting,quickdraw,real,sketch',
+        base_dir='D:/My_Codes/Federate-Learning/Data/DomainNet/'
```
Comment (kxzxvbk) [inline flzoo/domainnet/domainnet_fedavg_resnet_config.py:9]: what is the difference between data_path and base_dir?

File: flzoo/domainnet/domainnet_fedavg_resnet_config.py [hunk 26]
```diff
@@ -0,0 +1,35 @@
...
+        global_eps=500, 
+        batch_size=64, 
+        optimizer=dict(name='sgd', lr=0.01, momentum=0.5)
+    ),
+    model=dict(
+        name='resnet18',
+        input_channel=3,
+        class_number=10,
+    ),
+    client=dict(name='cross_domain_base_client', client_num=1),
```
Comment (kxzxvbk) [inline flzoo/domainnet/domainnet_fedavg_resnet_config.py:23]: Does ``client_num=1`` means the number of clients for each domain is 1? Clarify it and add more comments.

File: flzoo/domainnet/domainnet_fedavg_resnet_config.py [hunk 27]
```diff
@@ -0,0 +1,35 @@
...
+    server=dict(name='cross_domain_base_server'),
+    group=dict(name='cross_domain_base_group', aggregation_method='avg'),
+    # add group personal choice: Only aggregate parameters whose name does not contain the keyword "fc".
+        # aggregation_parameters=dict(name='except', keywords=['fc']
+    other=dict(test_freq=5, logging_path='./logging/cross_domain_domainnet_fedavg_resnet_iid_500')
```
Comment (kxzxvbk) [inline flzoo/domainnet/domainnet_fedavg_resnet_config.py:28]: Give a clearer logging_path name
